src/snmp_agent.c

```diff
@@ -0,0 +1,1478 @@
+                                 data_definition_t *dd, char *instance,
+                                 int oid_index) {
+  char buf[DATA_MAX_NAME_LEN];
+  format_name(buf, sizeof(buf), hostname_g, dd->plugin,
+              instance ? instance : dd->plugin_instance, dd->type,
+              dd->type_instance);
+  DEBUG(PLUGIN_NAME ": Identifier '%s'", buf);
+
+  value_t *values;
+  size_t values_num;
+  const data_set_t *ds = plugin_get_ds(dd->type);
+  if (ds == NULL) {
+    DEBUG(PLUGIN_NAME ": Data set not found for '%s' type", dd->type);
+    return SNMP_NOSUCHINSTANCE;
+  }
+
+  int ret = uc_get_value_by_name(buf, &values, &values_num);
+
+  if (ret != 0) {
+    ERROR(PLUGIN_NAME ": Failed to get value for '%s'", buf);
+    return SNMP_NOSUCHINSTANCE;
+  }
+
+  assert(ds->ds_num == values_num);
+  assert(oid_index < values_num);
+
+  char data[DATA_MAX_NAME_LEN];
+  size_t data_len = sizeof(data);
+  ret = snmp_agent_set_vardata(
+      data, &data_len, dd->oids[oid_index].type, dd->scale, dd->shift,
+      &values[oid_index], sizeof(values[oid_index]), ds->ds[oid_index].type);
+
+  sfree(values);
+
+  if (ret != 0) {
+    ERROR(PLUGIN_NAME ": Failed to convert '%s' value to snmp data", buf);
+    return SNMP_NOSUCHINSTANCE;
+  }
+
+  requests->requestvb->type = dd->oids[oid_index].type;
+  snmp_set_var_typed_value(requests->requestvb, requests->requestvb->type, data,
+                           data_len);
+
+  return SNMP_ERR_NOERROR;
+}
+
+static int
+snmp_agent_table_oid_handler(struct netsnmp_mib_handler_s *handler,
+                             struct netsnmp_handler_registration_s *reginfo,
+                             struct netsnmp_agent_request_info_s *reqinfo,
+                             struct netsnmp_request_info_s *requests) {
+
+  if (reqinfo->mode != MODE_GET && reqinfo->mode != MODE_GETNEXT) {
+    DEBUG(PLUGIN_NAME ": Not supported request mode (%d)", reqinfo->mode);
+    return SNMP_ERR_NOERROR;
+  }
+
+  pthread_mutex_lock(&g_agent->lock);
+
+  oid_t oid;
+  memcpy(oid.oid, requests->requestvb->name,
+         sizeof(oid.oid[0]) * requests->requestvb->name_length);
+  oid.oid_len = requests->requestvb->name_length;
+
+#if COLLECT_DEBUG
+  char oid_str[DATA_MAX_NAME_LEN];
+  snmp_agent_oid_to_string(oid_str, sizeof(oid_str), &oid);
+  DEBUG(PLUGIN_NAME ": Get request received for table OID '%s'", oid_str);
+#endif
+
+  for (llentry_t *te = llist_head(g_agent->tables); te != NULL; te = te->next) {
+    table_definition_t *td = te->value;
+
+    for (llentry_t *de = llist_head(td->columns); de != NULL; de = de->next) {
+      data_definition_t *dd = de->value;
+
+      for (int i = 0; i < dd->oids_len; i++) {
+        int ret = snmp_oid_ncompare(oid.oid, oid.oid_len, dd->oids[i].oid,
+                                    dd->oids[i].oid_len,
+                                    MIN(oid.oid_len, dd->oids[i].oid_len));
+        if (ret != 0)
+          continue;
+
+        if (!td->index_oid.oid_len) {
```

Comment:
You can do this right after initializing `td` and avoid two nested loops that will never be able to perform any useful action.